lte/gateway/python/magma/pipelined/rpc_servicer.py

```diff
@@ -788,6 +806,136 @@ def _ng_tunnel_update(self, pdr_entry: PDRRuleEntry, subscriber_id: str) -> bool
 
         return ret
 
+    def _ng_qer_update(self, request:SessionSet, pdr_entry: PDRRuleEntry
+                       ) -> Tuple[List[RuleModResult], List[RuleModResult]]:
+        enforcement_res = []
+        failed_policy_rules_results = []
+
+        subscriber_id = request.subscriber_id
+        session_version = request.session_version
+        ng_session_id = request.local_f_teid
+
+        # PDR is deleted with ActiveRules or DelActive rules recieved
+        if pdr_entry.pdr_state == PdrState.Value('REMOVE'):
+            self._ng_deactivate_qer_flows(subscriber_id, ng_session_id,
+                                          pdr_entry, session_version)
+        elif pdr_entry.pdr_state == PdrState.Value('IDLE'):
+            self._ng_inactivate_qer_flows(subscriber_id, pdr_entry, session_version)
+        #Install PDR rules
+        elif pdr_entry.pdr_state == PdrState.Value('INSTALL'):
+            if pdr_entry.del_qos_enforce_rule:
+                self._ng_deactivate_qer_flows(subscriber_id, ng_session_id,
+                                              pdr_entry, session_version)
+
+            enforcement_res = \
+                self._ng_activate_qer_flow(subscriber_id, ng_session_id,
+                                           pdr_entry, session_version)
+
+            failed_policy_rules_results = \
+                    _retrieve_failed_results(enforcement_res)
+
+        return failed_policy_rules_results
+
+    def _ng_inactivate_qer_flows(self, subscriber_id, pdr_entry, session_version):
+
+        ipv4 = convert_ipv4_str_to_ip_proto(pdr_entry.ue_ip_addr)
+        self._ng_update_version(subscriber_id, pdr_entry.add_qos_enforce_rule,
+                                ipv4, session_version)
+
+        if pdr_entry.add_qos_enforce_rule:
+            qos_enforce_rule = pdr_entry.add_qos_enforce_rule
+            rule_ids = [policy.rule.id for policy in qos_enforce_rule.policies]
+
+
+        if rule_ids:
+            self._enforcer_app.deactivate_rules(subscriber_id, ipv4,
+                                                rule_ids)
+
+    def _ng_activate_qer_flow(self, subscriber_id, ng_session_id,
+                              pdr_entry, session_version):
+
+        ipv4 = convert_ipv4_str_to_ip_proto(pdr_entry.ue_ip_addr)
```

Comment:
Okay should we just call the _install_flows_gx function here? I don't see any particular differences.

pull this block out into a common function `install_policy_enforcement` maybe?
```
        # Install rules in enforcement stats
        enforcement_stats_res = self._activate_rules_in_enforcement_stats(
                                         subscriber_id, qos_enforce_rule.msisdn,
                                         pdr_entry.local_f_teid,
                                         ipv4, qos_enforce_rule.apn_ambr,
                                         qos_enforce_rule.policies, ng_session_id)

        failed_policies_results = \
             _retrieve_failed_results(enforcement_stats_res)

        # Do not install any rules that failed to install in enforcement_stats.
        policy_rules = \
             _filter_failed_policies(qos_enforce_rule, failed_policies_results)

        enforcement_res = \
               self._activate_rules_in_enforcement(
                    subscriber_id, qos_enforce_rule.msisdn, pdr_entry.local_f_teid,
                    ipv4, qos_enforce_rule.apn_ambr,
                    policy_rules, ng_session_id)

        # Include the failed rules from enforcement_stats in the response.
        enforcement_res.policy_results.extend(
             failed_policies_results)

```

Reply:
Version calculation is only different between two functions.
Right now SMF is not sending version information for qos rule. We are using as session version for rule version.
Later we will do implementation for SMF side as similar to 4G.

Reply:
I really don't like us duplicating so much code here.  I don't see why can't you do

```
....

self._ng_update_version(subscriber_id, pdr_entry.add_qos_enforce_rule,
                                ipv4, session_version)
                                
self._install_flows_gx(qos_enforce_rule, ipv4)
 ```